src/Compilers/CSharp/Portable/FlowAnalysis/DataFlowPass.cs

```diff
@@ -903,43 +903,55 @@ protected void CheckAssigned(Symbol symbol, SyntaxNode node)
+                        ReportUnassignedIfOutsideLocalFunction(symbol, node, slot);
                     }
                 }
             }
         }
 
-        /// <param name="symbol">Symbol to variable that is unassigned.</param>
-        /// <param name="node">Syntax where read occurs.</param>
-        /// <param name="slotOpt">Optional slot where variable is located.</param>
-        /// <param name="skipIfUseBeforeDeclaration">
-        /// True if error reporting should consider the location where the
-        /// variable is declared (for instance, eliding errors about reading
-        /// variables that have not yet been declared).
-        /// </param>
-        private void ReportUnassigned(Symbol symbol, SyntaxNode node, int? slotOpt, bool skipIfUseBeforeDeclaration = true)
+        private void ReportUnassignedIfOutsideLocalFunction(Symbol symbol, SyntaxNode node, int slot, bool skipIfUseBeforeDeclaration = true)
         {
-            int slot = slotOpt ?? VariableSlot(symbol);
-            if (slot <= 0) return;
+            // If the symbol is captured by the nearest
+            // local function, record the read and skip the diagnostic
+            if ((object)GetNearestLocalFunctionOpt(currentMethodOrLambda) != null)
+            {
+                // If this is a constant, constants are always definitely assigned
+                // so we should skip reporting. This can happen in a local function
+                // where we use a constant before we actually visit its definition
+                // (since local function declarations are visited before other statements).
+                if (symbol is LocalSymbol local && local.IsConst)
+                {
+                    return;
+                }
+                else if (IsCapturedInLocalFunction(slot))
+                {
+                    RecordReadInLocalFunction(slot);
+                    return;
+                }
+            }
+
+            ReportUnassigned(symbol, node, slot, skipIfUseBeforeDeclaration);
+        }
+
+        /// <summary>
+        /// Report a given variable as not definitely assigned.  Once a variable has been so
+        /// reported, we suppress further reports of that variable.
+        /// </summary>
+        protected virtual void ReportUnassigned(Symbol symbol, SyntaxNode node, int slot, bool skipIfUseBeforeDeclaration)
+            => ReportUnassignedHelper(symbol, node, slot, skipIfUseBeforeDeclaration);
 
-            // If this is a constant, constants are always definitely assigned
-            // so we should skip reporting. This can happen in a local function
-            // where we use a constant before we actually visit its definition
-            // (since local function declarations are visited before other statements).
-            if (symbol.Kind == SymbolKind.Local && ((LocalSymbol)symbol).IsConst)
+        private void ReportUnassignedHelper(Symbol symbol, SyntaxNode node, int slot, bool skipIfUseBeforeDeclaration)
```

Comment:
>ReportUnassignedHelper [](start = 21, length = 22)

It doesn't look like we are taking any advantage from having this helper method. Please merge it with ReportUnassigned.